src/blockMonitor/blockCache.ts

```diff
@@ -99,110 +92,142 @@ export class BlockCache<TBlock extends IBlockStub> implements ReadOnlyBlockCache
+    // If so, add them and repeat with the next height (as some blocks might now have become complete)
+    private processPending(height: number) {
+        const blockHashesAtHeight = this.blockHashesByHeight.get(height) || new Set();
+        const blockHashesToAdd = [...blockHashesAtHeight].filter(h => this.pendingBlocksByHash.has(h));
+
+        for (const blockHash of blockHashesToAdd) {
+            const block = this.pendingBlocksByHash.get(blockHash)!;
+
+            // Remove block from pendingBlocksByHash, add to blocksByHash
+            this.blocksByHash.set(blockHash, block);
+            this.pendingBlocksByHash.delete(blockHash);
+
+            // Might need to update the maximum height
+            this.updateMaxHeightAndPrune(block.number);
         }
 
-        if (block.number < this.minHeight) {
-            // block too deep
-            return false;
+        if (blockHashesToAdd.length > 0) {
+            this.processPending(height + 1);
+        }
+    }
+
+    // If minHeight is increased after adding some blocks, some previously pending blocks should now be moved into blocksByHash.
+    // Since the process itself could (in rare circumstances) also increase minHeight, we check if this is the case and repeat the cycle.
+    private processPendingBlocksAtMinHeight() {
+        let prevMinHeight: number;
+        do {
+            prevMinHeight = this.minHeight;
+            this.processPending(this.minHeight);
+        } while (this.minHeight > prevMinHeight); // if the minHeight increased, run again
+    }
+
+    private updateMaxHeightAndPrune(newHeight: number) {
+        // If the maximum block height increased, we might have to prune some old info
+        if (this.mMaxHeight < newHeight) {
+            this.mMaxHeight = newHeight;
+            this.prune();
         }
-        return true;
     }
 
     /**
      * Adds `block`to the cache.
      * @param block
-     * @returns `true` if the block was added, `false` if the block was not added (because too deep or already in cache).
-     * @throws `ApplicationError` if the block cannot be added because its parent is not in cache.
+     * @returns `false` if the block was added (or was already present) as pending, `true` otherwise.
+     *      Note: it will return `true` even if the block was not actually added because already present, or because deeper than `maxDepth`.
      */
     public addBlock(block: Readonly<TBlock>): boolean {
-        // If the block's parent is above the minimum visible height, it needs to be added first
-        if (!this.canAddBlock(block)) {
-            throw new ApplicationError("Tried to add a block before its parent block.");
-        }
+        if (this.blocksByHash.has(block.hash)) return true; // block already added
+        if (block.number < this.minHeight) return true; // block already too deep, nothing to do
+
+        if (this.pendingBlocksByHash.has(block.hash)) return false; // block already pending
```

Comment:
No, the BlockCache immediately makes a pending block complete/attached as soon as possible. So if it's in "pending", at least one ancestor is still missing before it can be attached.